modules/widgets/instagram/instagram.php

```diff
@@ -0,0 +1,470 @@
+	 *
+	 * @param array $instance The widget instance (configuration options).
+	 */
+	public function form( $instance ) {
+		$instance = wp_parse_args( $instance, $this->defaults );
+
+		// If coming back to the widgets page from an action, expand this widget
+		if ( isset( $_GET['instagram_widget_id'] ) && $_GET['instagram_widget_id'] == $this->number ) {
+			echo '<script type="text/javascript">jQuery(document).ready(function($){ $(\'.widget[id$="wpcom_instagram_widget-' . esc_js( $this->number ) . '"] .widget-inside\').slideDown(\'fast\'); });</script>';
+		}
+
+		// If coming back from an OAuth authentication, validate and use the one in the URL
+		if ( isset( $_GET['instagram_widget_id'] ) && $_GET['instagram_widget_id'] == $this->number
+			&& ! empty( $_GET['instagram_widget'] ) && 'connection_verified' == $_GET['instagram_widget']
+			&& ! empty( $_GET['token_id'] ) && $instance['token_id'] !== (int) $_GET['token_id']  ) {
+			$instance['token_id'] = (int) $_GET['token_id'];
+
+			// check the nonce
+			$site = Jetpack_Options::get_option( 'id' );
+			$path = sprintf( '/sites/%s/instagram/verify-nonce/%s/%s', $site, $_GET['wpcom_nonce'], $_GET['blog'] );
+			$result = $this->wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+			$response_code = wp_remote_retrieve_response_code( $result );
+			if ( 200 !== $response_code ) {
+				do_action( 'wpcomsh_log', 'Instagram widget: failed to verify nonce: API returned code ' . $response_code );
+				return 'ERROR';
+			}
+
+			$this->update_widget_token_id( $instance['token_id'] );
+			$this->update_widget_token_legacy_status( false );
+		}
+		// If removing the widget's stored token ID
+		elseif ( $instance['token_id'] && isset( $_GET['instagram_widget_id'] ) && $_GET['instagram_widget_id'] == $this->number && ! empty( $_GET['instagram_widget'] ) && 'remove_token' == $_GET['instagram_widget'] ) {
+			if ( empty( $_GET['nonce'] ) || ! wp_verify_nonce( $_GET['nonce'], 'instagram-widget-remove-token-' . $this->number . '-' . $instance['token_id'] ) ) {
+				wp_die( __( 'Missing or invalid security nonce.', 'wpcom-instagram-widget' ) );
+			}
+			$site = Jetpack_Options::get_option( 'id' );
+			$path = sprintf( '/sites/%s/instagram/%s/remove', $site, $instance['token_id'] );
+			$result = $this->wpcom_json_api_request_as_blog( $path, 2, array( 'headers' => array( 'content-type' => 'application/json' ) ), null, 'wpcom' );
+
+			$response_code = wp_remote_retrieve_response_code( $result );
+
+			if ( 200 !== $response_code ) {
+				do_action( 'wpcomsh_log', 'Instagram widget: failed to remove keyring token: API returned code ' . $response_code );
+				return 'ERROR';
+			}
+
+			$instance['token_id'] = $this->defaults['token_id'];
+
+			$this->update_widget_token_id( $instance['token_id'] );
+			$this->update_widget_token_legacy_status( false );
+			?>
+			<script type="text/javascript">
```

Comment:
This works to remove the query params that cause errors on reloading after removing the existing token - works perfectly on Widgets page, but doesn't fire until the widget is opened again on customizer - but this may not actually be an issue - I tried enqueing this script instead but then it didn't fire at all.